# Exercise 3: Classification using MLP 

In this exercise student familiarizes the neural network classification problem.
The purpose is use MLP classification in Penguin  Dataset to classify different species just like in the Iris dataset. The dataset contains measurement of 3 different species of penguins coming from 3 islands in the Palmer Archipelago. These three classes are Adelie, Chinstrap, and Gentoo. If ‘Gentoo’ sounds familiar that is because Gentoo Linux is named after it! Also, these datasets contain culmen dimensions for each species. The culmen is the upper ridge of a bird’s bill. 

Columns in dataset:
- __species__ a factor denoting penguin species (Adélie, Chinstrap and Gentoo)
- __island__ a factor denoting island in Palmer Archipelago, Antarctica (Biscoe, Dream or Torgersen)
- __bill_length_mm__ a number denoting bill length (millimeters)
- __bill_depth_mm__ a number denoting bill depth (millimeters)
- __flipper_length_mm__ an integer denoting flipper length (millimeters)
- __body_mass_g__ an integer denoting body mass (grams)
- __sex__ a factor denoting penguin sex (female, male)

Scatter plots of features can be found [Github](https://github.com/allisonhorst/palmerpenguins)

More information about the dataset:
[Penguin dataset on Kaggle](https://www.kaggle.com/parulpandey/palmer-archipelago-antarctica-penguin-data/version/1)

Dataset can be found from Moodle: penguins_size.csv


In [1]:
# Import libraries and dataset
# Implementation comes this cell

# Import necessary libraries
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split

import tensorflow as tf
from tensorflow.keras import Input, Model
from tensorflow.keras.layers import Dense

# Load the dataset and display the first few rows
df_penguins = pd.read_csv('penguins_size.csv')
print(df_penguins.head())

  species     island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen              39.1             18.7              181.0   
1  Adelie  Torgersen              39.5             17.4              186.0   
2  Adelie  Torgersen              40.3             18.0              195.0   
3  Adelie  Torgersen               NaN              NaN                NaN   
4  Adelie  Torgersen              36.7             19.3              193.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  


In [9]:
# Data examination (You can do as you want)
# Implementation comes this cell

# Get basic information about the dataset
df_penguins.info() 

# Check for unique values in categorical columns
for col in ['species', 'island', 'sex']:
    print(col, df_penguins[col].unique())

# Treat the invalid '.' in sex as missing, then drop rows with any null
df_penguins['sex'] = df_penguins['sex'].replace('.', np.nan)

# Check for missing values and drop rows with any null values
print(df_penguins.isnull().sum())
df_penguins = df_penguins.dropna().reset_index(drop=True)
print(df_penguins.isnull().sum())

# Get the shape of the dataset after cleaning
print("Shape after cleaning:", df_penguins.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 333 entries, 0 to 332
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            333 non-null    object 
 1   island             333 non-null    object 
 2   culmen_length_mm   333 non-null    float64
 3   culmen_depth_mm    333 non-null    float64
 4   flipper_length_mm  333 non-null    float64
 5   body_mass_g        333 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.3+ KB
species ['Adelie' 'Chinstrap' 'Gentoo']
island ['Torgersen' 'Biscoe' 'Dream']
sex ['MALE' 'FEMALE']
species              0
island               0
culmen_length_mm     0
culmen_depth_mm      0
flipper_length_mm    0
body_mass_g          0
sex                  0
dtype: int64
species              0
island               0
culmen_length_mm     0
culmen_depth_mm      0
flipper_length_mm    0
body_mass_g          0

In [3]:
# Categorical varibles: Encode columns Species, Island and Sex
# species = Adélie, Chinstrap and Gentoo
# sex = MALE, FEMALA
# island = Torgersen, Biscoe and Dream

# Features = island, culmen_length_mm,culmen_depth_mm, flipper_length_mm, body_mass_g & sex
# Label = species

# Implementation comes this cell

# Group numeric and categorical features separately for preprocessing
categorical_cols    = ["island", "sex"]
numeric_cols    = ["culmen_length_mm", "culmen_depth_mm", "flipper_length_mm", "body_mass_g"]

# Convert categorical features to dummy variables
dummies_categorical_df = pd.get_dummies(df_penguins[categorical_cols], dtype=int)

# Scale all the numeric features using StandardScaler
scaler_x = StandardScaler()
scaled_numeric = scaler_x.fit_transform(df_penguins[numeric_cols]) # returns NumPy array

# Prepare features (X)
X_scaled = np.concatenate((scaled_numeric, dummies_categorical_df.to_numpy()), axis=1)

# Prepare target variable (y)
enc = LabelEncoder()
y_labeled = enc.fit_transform(df_penguins["species"])

In [4]:
# Dataset splitting
# training set 80%
# test set 20%
# Implementation comes this cell

X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y_labeled, test_size=0.2, random_state=42, stratify=y_labeled
)

In [5]:

# Define neural network using Keras Function API
# Input layer
# Hidden layers  
# Output layer 

# Implementation comes this cell


# Print the structure of model: model.summary()

In [6]:
# Compile with following parameters
# optimizer= adam 
# loss= sparse_categorical_crossentropy
# metrics= accuracy
# Implementation comes this cell


In [7]:
# Fit model with 
# epochs: test with different amout of epoch and write down the accurary
# Implementation comes this cell


In [ ]:
# Evaluate()
# Implementation comes this cell

n_features = X_train.shape[1]   # 9
n_classes = len(enc.classes_)   # 3

layer_nodes_combo_options = [[2], [4], [8], [16], [4, 2], [8, 4]]
epoch_options = [10, 50, 100, 200]
n_runs = 5   # train each config 5 times with different seeds

results_table = []

for layer_nodes_combo in layer_nodes_combo_options:
    for n_epochs in epoch_options:
        test_acc_runs = []   # this config's test accuracy across all runs

        for run in range(n_runs):
            tf.keras.utils.set_random_seed(run)   # seeds 0, 1, 2, 3, 4

            inputs = Input(shape=(n_features,))
            x = inputs
            for n_nodes in layer_nodes_combo:
                x = Dense(n_nodes, activation='relu')(x)
            outputs = Dense(n_classes, activation='softmax')(x)
            model = Model(inputs=inputs, outputs=outputs)
            model.compile(optimizer='adam',
                          loss='sparse_categorical_crossentropy',
                          metrics=['accuracy'])

            model.fit(X_train, y_train, epochs=n_epochs, verbose=0)
            test_loss, test_acc = model.evaluate(X_test, y_test, verbose=0)
            test_acc_runs.append(test_acc)

        avg_acc = np.mean(test_acc_runs)
        std_acc = np.std(test_acc_runs)

        results_table.append((layer_nodes_combo, n_epochs, avg_acc, std_acc))
        print(f"Layer/Nodes: {layer_nodes_combo}, Epochs: {n_epochs} → "
              f"Avg Test Accuracy: {avg_acc:.4f} (± {std_acc:.4f} across {n_runs} runs)")

Layer/Nodes: [2], Epochs: 10 → Avg Test Accuracy: 0.4418 (± 0.2344 across 5 runs)
Layer/Nodes: [2], Epochs: 50 → Avg Test Accuracy: 0.8179 (± 0.0466 across 5 runs)
Layer/Nodes: [2], Epochs: 100 → Avg Test Accuracy: 0.9701 (± 0.0163 across 5 runs)
Layer/Nodes: [2], Epochs: 200 → Avg Test Accuracy: 1.0000 (± 0.0000 across 5 runs)
Layer/Nodes: [4], Epochs: 10 → Avg Test Accuracy: 0.6896 (± 0.1158 across 5 runs)
Layer/Nodes: [4], Epochs: 50 → Avg Test Accuracy: 0.9642 (± 0.0202 across 5 runs)
Layer/Nodes: [4], Epochs: 100 → Avg Test Accuracy: 0.9970 (± 0.0060 across 5 runs)
Layer/Nodes: [4], Epochs: 200 → Avg Test Accuracy: 0.9970 (± 0.0060 across 5 runs)
Layer/Nodes: [8], Epochs: 10 → Avg Test Accuracy: 0.7015 (± 0.1280 across 5 runs)
Layer/Nodes: [8], Epochs: 50 → Avg Test Accuracy: 0.9881 (± 0.0174 across 5 runs)
Layer/Nodes: [8], Epochs: 100 → Avg Test Accuracy: 1.0000 (± 0.0000 across 5 runs)
Layer/Nodes: [8], Epochs: 200 → Avg Test Accuracy: 0.9940 (± 0.0073 across 5 runs)
Layer/Node

__Conclusion__

__NOTE: Conclusion is one evaluation criteria for the exercise, so pay attention to writing them__

## Write conclusion:
### How number of nodes in hidden layer affects the results?
More nodes made the network learn faster. With only 10 epochs, the network with 2 nodes got about 44% accuracy, but the network with 16 nodes got about 92%. The 16-node network reached 100% accuracy after only 50 epochs. 
A small network can still learn well, but it needs more training. The 2-node network also reached 100% accuracy, but only after 200 epochs. 
I also tried networks with two hidden layers. They did not work better. The [4, 2] network was the worst, because its second layer has only 2 nodes and the network must choose between 3 species. This makes a "bottleneck". The [8, 4] network was better, but still slower than one layer with 8 nodes. For this dataset, one hidden layer with 8–16 nodes was the best choice.

### How the number of epochs affects the results?
More epochs gave better accuracy for every network. With 10 epochs, all networks were too weak (underfitting), and the results changed a lot between runs. With 100–200 epochs, most networks reached 97–100% accuracy, and the results were much more stable. 
After the network reached 100%, more epochs did not help. Some results even dropped a little at 200 epochs, but the drop was very small (only a few wrong predictions), so it is probably just random variation.

### What you learn in this exercise?
I have learnt quite a lot in tis exercise as compare to previous exercise
  1. How to build a neural network for classification with the Keras Functional API
  2. Why data must be cleaned first: the dataset had missing values and one wrong value (".")
  that `isnull()` could not find
  3. Why text columns must be changed into numbers: dummy encoding for features and label
  encoding for the species
  4. Why scaling is important: the features had very different sizes, for example body mass
  in thousands and bill depth around 17
  5. Why `stratify` is useful when the classes are not equal in size
  6. That random starting weights change the results, so it is better to train each setup
  several times and use the average
  7. That a bigger or deeper network is not always better

### Did you have difficulties in the exercise?
Yes. At first, all my models gave exactly the same accuracy (0.4348), which looked strange.
The problem was that my notebook was still using old data with missing values, because I ran
cells in the wrong order. The missing values broke the training, and every model predicted
the same species. I fixed it by restarting the kernel and running all cells again from the top.
Now I always check the data shapes and missing values after splitting.

It was also a bit difficult to understand when to scale the data. The correct way is to
split the data first and then scale it, so the model does not learn anything from the test
data. In my main version, I scaled the data before splitting because it was simpler. I also
made a second version that scales after splitting, to compare. The effect is very small for
this dataset, but I understand that scaling after splitting is the better practice.

- Etc.

